src/Microsoft.DotNet.Tools.Repl.Csi/Program.cs

```diff
@@ -14,26 +19,148 @@ public static int Main(string[] args)
         {
             DebugHelper.HandleDebugSwitch(ref args);
 
-            var app = new CommandLineApplication();
+            var app = new CommandLineApplication(throwOnUnexpectedArg: false);
             app.Name = "dotnet repl csi";
             app.FullName = "C# REPL";
             app.Description = "C# REPL for the .NET platform";
             app.HelpOption("-h|--help");
+
             var script = app.Argument("<SCRIPT>", "The .csx file to run. Defaults to interactive mode.");
+            var framework = app.Option("-f|--framework <FRAMEWORK>", "Compile a specific framework", CommandOptionType.MultipleValue);
+            var configuration = app.Option("-c|--configuration <CONFIGURATION>", "Configuration under which to build", CommandOptionType.SingleValue);
+            var preserveTemporary = app.Option("-t|--preserve-temporary", "Preserve the temporary directory containing the compiled project.", CommandOptionType.NoValue);
+            var project = app.Option("-p|--project <PROJECT>", "The path to the project to run. Can be a path to a project.json or a project directory", CommandOptionType.SingleValue);
 
-            app.OnExecute(() => Run(script.Value));
+            app.OnExecute(() => Run(script.Value, framework.Values, configuration.Value(), preserveTemporary.HasValue(), project.Value(), app.RemainingArguments));
             return app.Execute(args);
         }
 
-        private static int Run(string scriptOpt)
+        private static ProjectContext GetProjectContext(IEnumerable<string> targetFrameworks, string projectPath)
+        {
+            var contexts = ProjectContext.CreateContextForEachFramework(projectPath);
+            var context = contexts.First();
+
+            if (targetFrameworks.Any())
+            {
+                var framework = NuGetFramework.Parse(targetFrameworks.First());
+                context = contexts.FirstOrDefault(c => c.TargetFramework.Equals(framework));
+            }
+
+            return context;
+        }
+
+        private static CommandResult CompileProject(ProjectContext projectContext, string configuration, out string tempOutputDir)
+        {
+            tempOutputDir = Path.Combine(projectContext.ProjectDirectory, "bin", ".dotnetrepl", Guid.NewGuid().ToString("N"));
+
+            Reporter.Output.WriteLine($"Compiling {projectContext.RootProject.Identity.Name.Yellow()} for {projectContext.TargetFramework.DotNetFrameworkName.Yellow()} to use with the {"C# REPL".Yellow()} environment.");
+
+            return Command.Create($"dotnet-compile", $"--output \"{tempOutputDir}\" --temp-output \"{tempOutputDir}\" --framework \"{projectContext.TargetFramework}\" --configuration \"{configuration}\" \"{projectContext.ProjectFile.ProjectDirectory}\"")
```

Comment:
@cston, They are the same, the code in `dotnet run` could probably be fixed up to use just `projectContext.ProjectDirectory` as well.

@KevinH-MS, The code is based on what `dotnet run` does for its temporary compilations. If you examine `dotnet build` or `dotnet compile`, you will see that `--temp-output` is actually the intermediate output folder and shouldn't cause conflicts over files on disk. I have added a comment to clarify.